File: backend/src/modules/points/point.service.js
const prisma = require('../../lib/prisma');
const { HttpError } = require('../../lib/httpError');
const historian = require('../../lib/historian');

async function getEquipmentContext(equipmentId) {
  const equipment = await prisma.equipment.findUnique({
    where: { id: equipmentId },
    include: { floor: true, building: true, site: true },
  });
  if (!equipment) {
    throw new HttpError(404, 'Equipment not found');
  }
  return equipment;
}

/**
 * @param {string} equipmentId
 * @param {{ skipSelfHeal?: boolean }} [options]
 */
async function listPointsByEquipment(equipmentId, options = {}) {
  await getEquipmentContext(equipmentId);
  const skipSelfHeal = options.skipSelfHeal === true;
  let rows = await prisma.point.findMany({
    where: { equipmentId },
    orderBy: { pointName: 'asc' },
  });
  if (rows.length === 0 && !skipSelfHeal) {
    const { syncSimCatalogBindingsForEquipmentId } = require('../../lib/simCatalogBindingSync');
    const synced = await syncSimCatalogBindingsForEquipmentId(equipmentId).catch(() => null);
    if (synced?.ok) {
      rows = await prisma.point.findMany({
        where: { equipmentId },
        orderBy: { pointName: 'asc' },
      });
    }
  }
  return rows;
}

async function createPoint(equipmentId, data) {
  const equipment = await getEquipmentContext(equipmentId);
  const siteId = equipment.siteId;
  const buildingId = equipment.buildingId;
  const floorId = equipment.floorId;

  const {
    pointName,
    pointCode,
    pointType,
    unit,
    writable,
    presentValue,
    status,
  } = data;

  if (!pointName || !pointCode || !pointType) {
    throw new HttpError(400, 'pointName, pointCode, and pointType are required');
  }

  const created = await prisma.point.create({
    data: {
      equipmentId,
      siteId,
      buildingId,
      floorId,
      pointName: String(pointName).trim(),
      pointCode: String(pointCode).trim(),
      pointType: String(pointType).trim(),
      unit: unit != null ? String(unit).trim() : null,
      writable: Boolean(writable),
      presentValue:
        presentValue != null ? String(presentValue) : null,
      ...(status ? { status } : {}),
    },
  });

  try {
    const alarmService = require('../alarms/alarm.service');
    await alarmService.syncAlarmDefinitionsAfterPointWrite(created);
    await alarmService.evaluateForPointIds([created.id]);
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('alarm sync after point create failed', e?.message || e);
  }

  return created;
}

async function getPointById(id) {
  const point = await prisma.point.findUnique({
    where: { id },
    include: {
      equipment: true,
      site: true,
      building: true,
      floor: true,
    },
  });
  if (!point) {
    throw new HttpError(404, 'Point not found');
  }
  return point;
}

async function updatePoint(id, data) {
  await getPointById(id);
  const allowed = [
    'pointName',
    'pointCode',
    'pointType',
    'unit',
    'writable',
    'presentValue',
    'status',
    'lastSeenAt',
    'commState',
  ];
  const update = {};
  for (const key of allowed) {
    if (data[key] !== undefined) {
      if (key === 'writable') {
        update[key] = Boolean(data[key]);
      } else if (key === 'presentValue' || key === 'unit') {
        update[key] =
          data[key] == null ? null : String(data[key]);
      } else if (key === 'lastSeenAt') {
        const v = data.lastSeenAt;
        update.lastSeenAt =
          v == null || v === '' ? null : v instanceof Date ? v : new Date(v);
      } else if (key === 'commState') {
        update.commState =
          data.commState == null || data.commState === ''
            ? null
            : String(data.commState).trim();
      } else {
        update[key] =
          typeof data[key] === 'string' ? data[key].trim() : data[key];
      }
    }
  }
  if (Object.keys(update).length === 0) {
    throw new HttpError(400, 'No fields to update');
  }
  const updated = await prisma.point.update({
    where: { id },
    data: update,
  });

  // Manual/engineering point updates are a real communication event for this
  // point, so they earn a historian sample the same as the SIM poll loop does
  // (LC-ARCH-001 D-010). This path is not inside a shared $transaction, so it
  // records immediately after the point write succeeds. It goes through the same
  // per-point sample-interval throttle as the SIM poll loop (historian.shouldRecordSample),
  // so a manual update landing in the same interval window as a SIM-driven write for the
  // same point can never create a second, near-duplicate row.
  if (update.presentValue !== undefined) {
    const sampleAt = update.lastSeenAt instanceof Date ? update.lastSeenAt : new Date();
    if (historian.shouldRecordSample(id, sampleAt.getTime())) {
      try {
        await historian.recordSample({
          pointId: id,
          value: update.presentValue,
          quality: historian.normalizeQuality(update.commState || updated.commState),
          timestamp: sampleAt,
        });
      } catch (e) {
        // eslint-disable-next-line no-console
        console.warn('historian sample after point update failed', e?.message || e);
      }
    }
  }

  try {
    const alarmService = require('../alarms/alarm.service');
    if (update.pointCode !== undefined) {
      await alarmService.syncAlarmDefinitionsAfterPointWrite(updated);
    }
    if (update.presentValue !== undefined || update.pointCode !== undefined) {
      await alarmService.evaluateForPointIds([id]);
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('alarm sync/evaluate after point update failed', e?.message || e);
  }

  return updated;
}

/**
 * Historian read path for the operator trend chart. Loads persisted samples for a set of
 * point ids within the requested range window, chronological ascending. Never fabricates
 * or backfills — a point with no samples in range simply returns an empty array for it.
 * @param {string[]} pointIds
 * @param {string} [range] - one of '1h' | '24h' | '7d' | '30d' (default '1h')
 * @returns {Promise<Record<string, Array<{ timestamp: string, value: string|null, quality: string }>>>}
 */
async function getHistoryForPointIds(pointIds, range) {
  const ids = Array.from(new Set((pointIds || []).map((id) => String(id || '').trim()).filter(Boolean)));
  const out = {};
  for (const id of ids) out[id] = [];
  if (ids.length === 0) return out;

  const windowMs = historian.windowMsForRange(range);
  const since = new Date(Date.now() - windowMs);

  const rows = await prisma.pointHistorySample.findMany({
    where: { pointId: { in: ids }, timestamp: { gte: since } },
    orderBy: { timestamp: 'asc' },
  });

  const byPoint = {};
  for (const id of ids) byPoint[id] = [];
  for (const row of rows) {
    byPoint[row.pointId].push({
      timestamp: row.timestamp.toISOString(),
      value: row.value,
      quality: row.quality,
    });
  }
  // Long ranges (7d/30d) at a fast poll rate can carry tens of thousands of real rows per point —
  // decimate each series independently to a safe maximum before it goes over the wire. This drops
  // rows, never values: every kept row is a real recorded sample, and the most recent sample is
  // always kept so the chart's "now" edge is never stale.
  for (const id of ids) out[id] = historian.decimateSamples(byPoint[id]);
  return out;
}

module.exports = {
  listPointsByEquipment,
  createPoint,
  getPointById,
  updatePoint,
  getHistoryForPointIds,
};

File: server/src/main/java/com/legioncontrols/server/sites/PointService.java
package com.legioncontrols.server.sites;

import com.legioncontrols.server.alarms.AlarmEvaluationService;
import com.legioncontrols.server.common.ApiException;
import com.legioncontrols.server.historian.HistorianService;
import com.legioncontrols.server.jooq.generated.enums.Entitystatus;
import com.legioncontrols.server.jooq.generated.tables.records.EquipmentRecord;
import com.legioncontrols.server.jooq.generated.tables.records.PointRecord;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import org.jooq.DSLContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import static com.legioncontrols.server.jooq.generated.tables.Equipment.EQUIPMENT;
import static com.legioncontrols.server.jooq.generated.tables.Point.POINT;

/**
 * Java equivalent of backend/src/modules/points/point.service.js. The Node version's
 * zero-points self-heal (lazily calling syncSimCatalogBindingsForEquipmentId) is a SIM-catalog/
 * BACnet-adjacent dev convenience intentionally NOT ported here — same category of deferred
 * convenience as the release-deploy SIM catalog resync (see ReleaseService); Express remains the
 * source of truth for that specific self-heal path until BACnet/SIM commissioning migrates.
 */
@Service
public class PointService {

    private static final Logger log = LoggerFactory.getLogger(PointService.class);

    private final DSLContext dsl;
    private final HistorianService historianService;
    private final AlarmEvaluationService alarmEvaluationService;

    public PointService(DSLContext dsl, HistorianService historianService, AlarmEvaluationService alarmEvaluationService) {
        this.dsl = dsl;
        this.historianService = historianService;
        this.alarmEvaluationService = alarmEvaluationService;
    }

    private EquipmentRecord getEquipmentContext(String equipmentId) {
        EquipmentRecord equipment = dsl.selectFrom(EQUIPMENT).where(EQUIPMENT.ID.eq(equipmentId)).fetchOne();
        if (equipment == null) {
            throw ApiException.notFound("Equipment not found");
        }
        return equipment;
    }

    public List<PointDto> listPointsByEquipment(String equipmentId) {
        getEquipmentContext(equipmentId);
        return dsl.selectFrom(POINT).where(POINT.EQUIPMENTID.eq(equipmentId)).orderBy(POINT.POINTNAME.asc())
            .fetch().stream().map(PointDto::from).toList();
    }

    public record CreatePointRequest(String pointName, String pointCode, String pointType, String unit,
                                      Boolean writable, String presentValue, String status) {
    }

    public PointRecord createPoint(String equipmentId, CreatePointRequest req) {
        EquipmentRecord equipment = getEquipmentContext(equipmentId);
        if (isBlank(req.pointName()) || isBlank(req.pointCode()) || isBlank(req.pointType())) {
            throw ApiException.badRequest("pointName, pointCode, and pointType are required");
        }

        PointRecord record = dsl.newRecord(POINT);
        record.setId(UUID.randomUUID().toString());
        record.setEquipmentid(equipmentId);
        record.setSiteid(equipment.getSiteid());
        record.setBuildingid(equipment.getBuildingid());
        record.setFloorid(equipment.getFloorid());
        record.setPointname(req.pointName().trim());
        record.setPointcode(req.pointCode().trim());
        record.setPointtype(req.pointType().trim());
        record.setUnit(req.unit() != null ? req.unit().trim() : null);
        record.setWritable(Boolean.TRUE.equals(req.writable()));
        record.setPresentvalue(req.presentValue());
        if (req.status() != null) record.setStatus(Entitystatus.valueOf(req.status().toUpperCase()));
        record.setCreatedat(LocalDateTime.now());
        record.setUpdatedat(LocalDateTime.now());
        record.insert();
        record.refresh();

        try {
            alarmEvaluationService.syncAlarmDefinitionsAfterPointWrite(record);
            alarmEvaluationService.evaluateForPointIds(List.of(record.getId()));
        } catch (Exception e) {
            log.warn("alarm sync after point create failed: {}", e.getMessage());
        }

        return record;
    }

    public PointRecord getPointRecordById(String id) {
        PointRecord point = dsl.selectFrom(POINT).where(POINT.ID.eq(id)).fetchOne();
        if (point == null) {
            throw ApiException.notFound("Point not found");
        }
        return point;
    }

    /**
     * Takes the raw request body as a JsonNode so PATCH semantics — "field omitted" vs. "field
     * explicitly null" — can be distinguished via {@code JsonNode.has(field)}, matching the Node
     * controller's {@code data[key] !== undefined} checks exactly (see AlarmDefinitionService's
     * updateDefinition for the same pattern and rationale).
     */
    public PointRecord updatePoint(String id, tools.jackson.databind.JsonNode body) {
        PointRecord existing = getPointRecordById(id);
        boolean changed = false;
        boolean pointCodeChanged = false;
        boolean presentValueSet = false;
        if (body.has("pointName")) { existing.setPointname(body.get("pointName").asText().trim()); changed = true; }
        if (body.has("pointCode")) { existing.setPointcode(body.get("pointCode").asText().trim()); changed = true; pointCodeChanged = true; }
        if (body.has("pointType")) { existing.setPointtype(body.get("pointType").asText().trim()); changed = true; }
        if (body.has("unit")) { existing.setUnit(body.get("unit").isNull() ? null : body.get("unit").asText()); changed = true; }
        if (body.has("writable")) { existing.setWritable(body.get("writable").asBoolean()); changed = true; }
        if (body.has("presentValue")) {
            existing.setPresentvalue(body.get("presentValue").isNull() ? null : body.get("presentValue").asText());
            changed = true;
            presentValueSet = true;
        }
        if (body.has("status")) { existing.setStatus(Entitystatus.valueOf(body.get("status").asText().toUpperCase())); changed = true; }
        LocalDateTime lastSeenAt = null;
        boolean lastSeenAtSet = body.has("lastSeenAt");
        if (lastSeenAtSet) {
            var node = body.get("lastSeenAt");
            lastSeenAt = (node == null || node.isNull() || node.asText("").isBlank())
                ? null
                : java.time.OffsetDateTime.parse(node.asText()).toLocalDateTime();
            existing.setLastseenat(lastSeenAt);
            changed = true;
        }
        if (body.has("commState")) {
            var node = body.get("commState");
            existing.setCommstate(node.isNull() || node.asText("").isBlank() ? null : node.asText().trim());
            changed = true;
        }
        if (!changed) {
            throw ApiException.badRequest("No fields to update");
        }
        existing.setUpdatedat(LocalDateTime.now());
        existing.update();

        // A manual/engineering point update is a real communication event for this point, so it
        // earns a historian sample the same as the SIM poll loop does — through the same
        // per-point sample-interval throttle, so a manual update in the same window as a SIM write
        // can never create a near-duplicate row.
        if (presentValueSet) {
            LocalDateTime sampleAt = lastSeenAtSet && lastSeenAt != null ? lastSeenAt : LocalDateTime.now();
            long sampleAtMs = sampleAt.toInstant(java.time.ZoneOffset.UTC).toEpochMilli();
            if (historianService.shouldRecordSample(id, sampleAtMs)) {
                try {
                    String quality = HistorianService.normalizeQuality(existing.getCommstate());
                    historianService.recordSample(new HistorianService.SampleEntry(id, existing.getPresentvalue(), quality, sampleAt));
                } catch (Exception e) {
                    log.warn("historian sample after point update failed: {}", e.getMessage());
                }
            }
        }

        try {
            if (pointCodeChanged) {
                alarmEvaluationService.syncAlarmDefinitionsAfterPointWrite(existing);
            }
            if (presentValueSet || pointCodeChanged) {
                alarmEvaluationService.evaluateForPointIds(List.of(id));
            }
        } catch (Exception e) {
            log.warn("alarm sync/evaluate after point update failed: {}", e.getMessage());
        }

        return existing;
    }

    public record HistoryResult(String range, java.util.Map<String, List<HistorianService.HistorySamplePoint>> samples) {
    }

    public HistoryResult getHistoryForPointIds(List<String> pointIds, String range) {
        String effectiveRange = range != null && !range.isBlank() ? range.trim() : "1h";
        return new HistoryResult(effectiveRange, historianService.getHistoryForPointIds(pointIds, effectiveRange));
    }

    private static boolean isBlank(String s) {
        return s == null || s.isBlank();
    }
}
